src/topology/instances/add_circle.lean

```diff
@@ -280,6 +285,72 @@ begin
     simpa [zero_add] using (equiv_Ico p 0 u).2.2, },
 end
 
+lemma add_order_of_eq_pos_iff {u : add_circle p} {n : ℕ} (h : 0 < n) :
+  add_order_of u = n ↔ ∃ m < n, gcd m n = 1 ∧ ↑(↑m / ↑n * p) = u :=
+begin
+  refine ⟨λ hu, _, _⟩,
+  { rw ← hu at h,
+    obtain ⟨m, h₀, h₁, h₂⟩ := exists_gcd_eq_one_of_is_of_fin_add_order (add_order_of_pos_iff.mp h),
+    refine ⟨m, _, _, _⟩;
+    rwa ← hu, },
+  { rintros ⟨m, h₀, h₁, rfl⟩,
+    exact add_order_of_div_of_gcd_eq_one h h₁, },
+end
+
+variables (p)
+
+/-- The natural bijection between points of order `n` and natural numbers less than and coprime to
+`n`. The inverse of the map sends `m ↦ (m/n * p : add_circle p)` where `m` is coprime to `n` and
+satisfies `0 ≤ m < n`. -/
+def set_add_order_of_equiv {n : ℕ} (hn : 0 < n) :
+  {u : add_circle p | add_order_of u = n} ≃ {m | m < n ∧ gcd m n = 1} :=
+{ to_fun := λ u, by
+  { let h := (add_order_of_eq_pos_iff hn).mp u.property,
+    exact ⟨classical.some h, classical.some (classical.some_spec h),
+      (classical.some_spec (classical.some_spec h)).1⟩, },
+  inv_fun := λ m, ⟨↑((m : 𝕜) / n * p), add_order_of_div_of_gcd_eq_one hn (m.property.2)⟩,
+  left_inv := λ u, subtype.ext
+    (classical.some_spec (classical.some_spec $ (add_order_of_eq_pos_iff hn).mp u.2)).2,
+  right_inv :=
+  begin
+    rintros ⟨m, hm₁, hm₂⟩,
+    let u : {u : add_circle p | add_order_of u = n} :=
+      ⟨↑((m : 𝕜) / n * p), add_order_of_div_of_gcd_eq_one hn hm₂⟩,
+    let h := (add_order_of_eq_pos_iff hn).mp u.property,
+    ext,
+    let m' := classical.some h,
+    change m' = m,
+    obtain ⟨h₁ : m' < n, h₂ : gcd m' n = 1, h₃ : quotient_add_group.mk ((m' : 𝕜) / n * p) =
+      quotient_add_group.mk ((m : 𝕜) / n * p)⟩ := classical.some_spec h,
+    replace h₃ := congr_arg (coe : Ico 0 (0 + p) → 𝕜) (congr_arg (equiv_Ico p 0) h₃),
+    simpa only [coe_equiv_Ico_mk_apply, mul_left_inj' hp.out.ne', mul_div_cancel _ hp.out.ne',
+      int.fract_div_nat_cast_eq_div_nat_cast_mod,
+      div_left_inj' (nat.cast_ne_zero.mpr hn.ne' : (n : 𝕜) ≠ 0), nat.cast_inj,
+      (nat.mod_eq_iff_lt hn.ne').mpr hm₁, (nat.mod_eq_iff_lt hn.ne').mpr h₁] using h₃,
+  end }
+
+@[simp] lemma card_add_order_of_eq_totient {n : ℕ} :
+  nat.card {u : add_circle p // add_order_of u = n} = n.totient :=
+begin
+  rcases n.eq_zero_or_pos with rfl | hn,
+  { simp only [nat.totient_zero, add_order_of_eq_zero_iff],
+    rcases em (∃ (u : add_circle p), ¬ is_of_fin_add_order u) with ⟨u, hu⟩ | h,
+    { haveI : infinite {u : add_circle p // ¬is_of_fin_add_order u},
+      { erw infinite_coe_iff,
+        exact infinite_not_is_of_fin_add_order hu, },
+      exact nat.card_eq_zero_of_infinite, },
+    { haveI : is_empty {u : add_circle p // ¬is_of_fin_add_order u}, { simpa using h, },
+      exact nat.card_of_is_empty, }, },
```

Comment:
I am not really satisfied by this proof: the set of points with infinite order is infinite, but you are (cleverly!) not proving this. I would be more happy with a proof that it is indeed infinite (unless this is too painful -- but the set of points with finite order is countable from what you prove, and the circle is not countable, so if we have enough API this should hopefully not be hard).

Reply:
The reason for this slightly odd proof is that I am still working over a linearly-ordered field, which might be `ℚ`, at this point. So there might really be no points of infinite order!

The statement is true because `nat.card` takes the same value on empty and infinite sets (which I think is the point that @alreadydone was making [above](https://github.com/leanprover-community/mathlib/pull/17986#discussion_r1057351748)).

Of course I'm happy to adjust as you see fit. I could add an independent lemma below proving that there are (uncountably-many) points of infinite order over `ℝ` if you like?

Reply:
Ah, I hadn't noticed you were on an arbitrary field. Then it's perfect as is. Thanks!